### ISSUE TITLE
async replication: move applier to iproto thread, aka parallel applier

### ISSUE THREAD

**Issue by kostja** [tarantool/tarantool#1254]
There is a user complain that our asynchronous replication never gets above 60K RPS since the applier is always running in a single fiber, so there is no group commit on replica. To enable group commit, we need to execute replication requests concurrently in multiple fibers, in the same manner this happens on the master.

There are no caveats for memtx, since it's going to sequentially execute the transactions anyway, but there are possible caveats for phia, which may require some form of internal locking to avoid applier conflicts.

**Comment by rtsisyk** [comment]
The first version was reviewed today.

**Comment by rtsisyk** [comment]
Postponed, blocked by #1612

**Comment by rtsisyk** [comment]
=> 1.8.0

**Comment by rtsisyk** [comment]
Blocked by #2798

**Comment by knazarov** [comment]
It looks like the changes haven't been merged. See: https://github.com/tarantool/tarantool/blob/f00945a185759b13314bd3bd8a778f1a3b11a68e/src/box/applier.cc#L626

**Comment by karlovnv** [comment]
In case of 50k ops: 25k inserts + 25k replaces per second we have:
**On master:**
```
tx: 44%
iproto: 13%
relay: 83%
wal: 35%
```

**On replica:**
```
tx: 75%
iproto: 0%
relay: 0%
wal: 35%
```

**Comment by kostja** [comment]
Which branch?

**Comment by karlovnv** [comment]
> Which branch?

2.2

An additional info:



  | TX (system), % | TX (user), % | TX (sum), %
-- | -- | -- | --
MASTER | 5 | 40,6 | **45,6**
REPLICA | **47,6** | 28,4 | **76**

Take a look at TX (system) on replica. It seems that too much time was spent in system space

**Comment by kostja** [comment]
@karlovnv  
Could you take a similar profile with 1.10? Frankly I don't trust 2.2, I don't know what it could be.

**Comment by karlovnv** [comment]
With 1.10 I see the same thing, but it cannot handle a such number of TPS with WAL enabled.

**Comment by kostja** [comment]
OK, glad to see there are no regressions. The improvement in CPU usage comes from write/read buffering done by Georgy in scope of parallel applier work. The CPU time on master will be reduced even more once Tarantool gets in-memory relay (which Cyrill needs to finish)

**Comment by karlovnv** [comment]
@kostja as you can see, there is no degradation. On the contrary: applier in 2.2 works much better than in 1.10 ( 2.2 replica can handle 50k ops, but 1.10 only 30k). 

As far as I know 2.2 includes parallel applier patches, as writing WAL does not block the main thread.

The main problem is that the applier performs all the CPU consuming stuff like parsing msgpack, working with socket, doing sys calls and so on in the TX thread. This should be moved to another thread (like iproto).

**Comment by kyukhin** [comment]
Superceeded by #6329
